Support to cache profiles created via `ProfileMapping`

pankajastro opened the pull request:
## Description

Add DBT profile caching mechanism.

1. Introduced env `enable_cache_profile` to enable or disable profile caching. This will be enabled only if global `enable_cache` is enabled. 
2. Users can set the env `profile_cache_dir_name`. This will be the name of a sub-dir inside `cache_dir` where cached profiles will be stored. This is optional, and the default name is `profile` 
3. Example Path for versioned profile: `{cache_dir}/{profile_cache_dir}/592906f650558ce1dadb75fcce84a2ec09e444441e6af6069f19204d59fe428b/profiles.yml`
4. Implemented profile mapping hashing: first, the profile is serialized using pickle. Then, the profile_name and target_name are appended before hashing the data using the SHA-256 algorithm

**Perf test result:**
In local dev env with command
```
AIRFLOW_HOME=`pwd` AIRFLOW_CONN_EXAMPLE_CONN="postgres://postgres:postgres@0.0.0.0:5432/postgres"  AIRFLOW_HOME=`pwd` AIRFLOW__CORE__DAGBAG_IMPORT_TIMEOUT=20000 AIRFLOW__CORE__DAG_FILE_PROCESSOR_TIMEOUT=20000 hatch run tests.py3.10-2.8:test-performance
```

NUM_MODELS=100
- TIME=167.45248413085938 (with profile cache enabled)
- TIME=173.94845390319824 (with profile cache disabled)

NUM_MODELS=200
- TIME=376.2585120201111 (with profile cache enabled)
- TIME=418.14210200309753 (with profile cache disabled)

## Related Issue(s)

Closes: #925
Closes: #647

## Breaking Change?

No

## Checklist

- [ ] I have made corresponding changes to the documentation (if required)
- [ ] I have added tests that prove my fix is effective or that my feature works

tatiana reviewed (commented):
Hi @pankajastro , thanks a lot for advancing on this!

Some questions/comments on the overall feature:
1. So we are consistent with other enabling cache mechanism (e.g. #1014), please, rename the config to `enable_cache_profile`
2. Why are we forcing users to defined the cache path, differently from #904, where we dynamically define a path - and we give the possibility for users to override it with the `cache_dir` config, if they want to. We could perhaps allow users to force a specific profile path if they want to, by having `cache_dir_profile` - but it would be great if this was optional.
3. What does the `metadata.json` relate to? A dbt profile, an Airflow connection, a `DbtDag`, or to a `DbtTaskGroup`, to the Airflow deployment..? The name is very generic and the scope is unclear. How are we handling scenarios where concurrent tasks/DAGs would be writing at the same time to this file?
4. How sensitive is the identifier of the cache (sha256) to changes in the Airflow connection being used?

tatiana commented on cosmos/config.py:
```diff
@@ -267,35 +299,40 @@ def ensure_profile(
-                # write profile_contents to desired_profile_path using yaml library
-                desired_profile_path.write_text(profile_contents)
-                yield desired_profile_path, env_vars
+            if enable_profile_cache:
```
This should be an end before the global setting `enable_cache` and the specific profile cache

pankajastro replied:
do you mean we should also check if `enable_cache` is also enable or something more?

tatiana commented on cosmos/config.py:
```diff
@@ -257,6 +259,36 @@ def validate_profiles_yml(self) -> None:
+        corresponding profile at given path.
+        """
+        # TODO: Fix type
+        profile_version = self.profile_mapping.version(use_mock_values)  # type: ignore[union-attr]
```
```suggestion
        current_profile_version = self.profile_mapping.version(use_mock_values)  # type: ignore[union-attr]
```

tatiana commented on cosmos/cache.py:
```diff
@@ -171,3 +173,64 @@ def _copy_partial_parse_to_project(partial_parse_filepath: Path, project_path: P
+    """
+    Check if profile version and correspond dbt profile yml exit in metadata.json and return the yml path or None
+    """
+    if not dbt_profile_cache_path:
```
This should be optional. We should not expect users to have to set this path - we already have a default `cache_dir`, why not reuse it?

pankajastro replied:
We can make it optional, I wanted to keep the profile caching-related files in one place. Shall we keep the profile related files in cache_dir / profile ?

tatiana commented on cosmos/config.py:
```diff
@@ -257,6 +259,36 @@ def validate_profiles_yml(self) -> None:
+
+                Path(Path(dbt_profile_cache_path) / profile_version).mkdir(parents=True, exist_ok=True)
+                logger.info("Profile not found in cache storing and using profile: %s.", profile_path)
+                profile_path.write_text(profile_contents)
```
With the current approach, we'd be storing all the caches for all the versions of profiles - regardless if they exist or not. Ideally, we would only keep the relevant cache to the latest definition of a specific profile, overriding it if the parameters changed.

pankajastro replied:
If the cache exists, i.e hash for the requested profile object exists in the cache path, then we are not rewriting it. 

> Ideally, we would only keep the relevant cache to the latest definition of a specific profile, overriding it if the parameters changed.

Could you please elaborate a bit more? Are you suggesting to keep only one profile cached per profile type? 

 was thinking that it makes sense to store in cache when object parameters change. For example, caching two profiles based on the object below, because the conn_id parameter has changed.

```
profile_mapping=PostgresUserPasswordProfileMapping(
        conn_id="airflow_db",
        profile_args={"schema": "public"},
        disable_event_tracking=True,
    ),
```

```
profile_mapping=PostgresUserPasswordProfileMapping(
        conn_id="airflow_db_1",
        profile_args={"schema": "public"},
        disable_event_tracking=True,
    ),
```

tatiana commented on cosmos/config.py:
```diff
@@ -257,6 +259,36 @@ def validate_profiles_yml(self) -> None:
         if self.profiles_yml_filepath and not Path(self.profiles_yml_filepath).exists():
             raise CosmosValueError(f"The file {self.profiles_yml_filepath} does not exist.")
 
+    def _handle_profile_caching(self, desired_profile_path: Path | None = None, use_mock_values: bool = False) -> Path:
```
Could we find a name that represents better what this method does? Ultimately, it returns the profile path.
And the docstring could also give a brief description on what it does, before detailing the how, something along the lines:
`Return previously cached profile path or calculate and store the profile content into a file and return its path`

pankajastro commented:
> How sensitive is the identifier of the cache (sha256) to changes in the Airflow connection being used?

hashing connection is still a TODO item and we need to pass the actual value of a connection in the hashing input

pankajastro commented:
> What does the metadata.json relate to? A dbt profile, an Airflow connection, a DbtDag, or to a DbtTaskGroup, to the Airflow deployment..?

A JSON file that stores key-value pairs, where each key is a hash for a Profile Mapping object and the value is the corresponding path to the DBT profile YAML file. I'm using this file to check if we already have profile YAML for this profile object hash 

> The name is very generic and the scope is unclear.

I'm. ok to change the name.

>  How are we handling scenarios where concurrent tasks/DAGs would be writing at the same time to this file?

I haven't considered the concurrent case, do we need locking here?

pankajastro commented:
> Why are we forcing users to defined the cache path, differently from https://github.com/astronomer/astronomer-cosmos/pull/904, where we dynamically define a path - and we give the possibility for users to override it with the cache_dir config, if they want to. We could perhaps allow users to force a specific profile path if they want to, by having cache_dir_profile - but it would be great if this was optional.

We can make it optional, I wanted to keep the profile caching-related files in one place. Shall we keep the profile related files in `cache_dir / profile` ?

pankajastro commented:
Improved the below items after our first discussion and review

- Added environment variable enable_profile_cache to toggle profile caching on or off.
- Introduced environment variable profile_cache_dir, which will be created within cache_dir. The default name is 'profile', and it will store the versioned DBT profile. 
- Example Path for versioned provifile: {cache_dir}/{profile_cache_dir}/version/profiles.yml
- Implemented profile mapping hashing: first, the profile is serialized using pickle. Then, the profile_name and target_name are appended before hashing the data using the SHA-256 algorithm
- Removed metadata.json because now I'm dealing with `desired_profile_path` variable
- Added a check to enable_profile_cache only if global caching is enabled

tatiana commented on docs/configuration/cosmos-conf.rst:
```diff
@@ -54,6 +54,23 @@ This page lists all available Airflow configurations that affect ``astronomer-co
     - Default: ``None``
     - Environment Variable: ``AIRFLOW__COSMOS__DBT_DOCS_CONN_ID``
 
+.. _enable_profile_cache:
```
So we are consistent with naming (e.g. #1014), let's use: `enable_cache_profile`. This way, all configs related to enabling cache share the same prefix.

Also, let's set this by default - so we are consistent with pre-existing cache.

tatiana commented on docs/configuration/cosmos-conf.rst:
```diff
@@ -54,6 +54,23 @@ This page lists all available Airflow configurations that affect ``astronomer-co
+    - Default: ``False``
+    - Environment Variable: ``AIRFLOW__COSMOS__ENABLE_PROFILE_CACHING``
+
+.. _profile_cache_path:
+
+`profile_cache_dir`_:
+    Folder path to store the DBT cached profile and related metadata
+
+    - Default: ``profile``
+    - Environment Variable: ``AIRFLOW__COSMOS__PROFILE_CACHE_DIR``
```
I believe we should not have this setting. Why don't we create a subdirectory in the existing `cache_dir` and avoid forcing users to have to manually add another configuration?

pankajastro replied:
This is optional and we are creating a sub dir default name  `profile` in `cache_dir` see https://github.com/astronomer/astronomer-cosmos/pull/1046/files#diff-16ea4e2752796e174ffdf472c3bd2ca64e4e7c62ab586e169f2631fc9bb7e564R182

pankajastro commented on tests/dbt/test_graph.py:
```diff
@@ -429,7 +429,7 @@ def test_load_via_dbt_ls_does_not_create_target_logs_in_original_folder(
     used_cwd = Path(mock_popen.call_args[0][0][-5])
     assert used_cwd != project_config.dbt_project_path
-    assert not used_cwd.exists()
+    # assert not used_cwd.exists()
```
This check is failing at the moment

pankajastro commented on docs/configuration/cosmos-conf.rst:
```diff
@@ -70,6 +70,23 @@ This page lists all available Airflow configurations that affect ``astronomer-co
+
+.. _profile_cache_dir:
+
+`profile_cache_dir`_:
```
```suggestion
`profile_cache_dir_name`_:
```

tatiana reviewed (commented):
HI @pankajastro thanks for addressing all the feedback!

Once you are able to uncomment that assert in the test and rename the config, I believe we're good to go! 🎉 

All the comments/feedback were addressed. Following up on:
> I haven't considered the concurrent case, do we need locking here?

I believe once we removed the `metadata.json` file, we're in a much safer position, since we reduce significantly concurrent writes to the same file. I suggest we go with the current solution and we review if users report any issues.

jlaneve commented on cosmos/profiles/base.py:
```diff
@@ -82,6 +84,23 @@ def __init__(
+            profile = self.profile
+        profile["profile_name"] = profile_name
+        profile["target_name"] = target_name
+        hash_object = hashlib.sha256(pickle.dumps(profile))
```
do we need to pickle here? or can we just json serialize/deserialize?

pankajastro replied:
I can opt for another serializer like `json.dumps` based on the complexity of the profile object. However, since pickle is already installed and works well to serialize complex objects, I'm using it here. I'm curious if there are any potential drawbacks to using pickle in this context?

tatiana replied:
Following up, @pankajastro , what we agreed:
1. It is worth changing from pickle to `json.dumps(sort_keys=True)
 for two reasons: the value needs to be deterministic and it must be safe to store it in a remote object store, something that's planned for Cosmos 1.6.
2. Add a test to validate that a given profile always has the same cache version. You may need to set a different value for Linux and MacOS.

pankajastro commented on tests/dbt/test_graph.py:
```diff
@@ -429,7 +432,10 @@ def test_load_via_dbt_ls_does_not_create_target_logs_in_original_folder(
-    assert not used_cwd.exists()
+    # When the cache profile is enabled, the profile is created in {cache_dir}/profile/{version}/profiles.yml
+    # rather than in a temporary file. This ensures it is persisted for future use and remains uncleared.
+    if not enable_cache_profile:
+        assert not used_cwd.exists()
```
@tatiana, `used_cwd` pointing to the dir where profiles.yml exists. When cache profile is enabled, it is created inside `cache_dir/profile/version`, ensuring persistence. If profile cache is enabled, does not it make sense, or I'm still missing something?

CI log:
```
FAILED tests/dbt/test_graph.py::test_load_via_dbt_ls_does_not_create_target_logs_in_original_folder[True] - AssertionError: assert not True
 +  where True = <bound method Path.exists of PosixPath('/tmp/cosmos/profile/470f1452ea88562203ec89a49bab7e2c6043ef3b7f93e61ce30772e24aa38289')>()
 +    where <bound method Path.exists of PosixPath('/tmp/cosmos/profile/470f1452ea88562203ec89a49bab7e2c6043ef3b7f93e61ce30772e24aa38289')> = PosixPath('/tmp/cosmos/profile/470f1452ea88562203ec89a49bab7e2c6043ef3b7f93e61ce30772e24aa38289').exists
```
Ref CI: https://github.com/astronomer/astronomer-cosmos/actions/runs/9679063910/job/26704228626

tatiana commented on tests/dbt/test_graph.py:
```diff
@@ -429,7 +429,7 @@ def test_load_via_dbt_ls_does_not_create_target_logs_in_original_folder(
 
     used_cwd = Path(mock_popen.call_args[0][0][-5])
```
@pankajastro could you confirm the value of `used_cwd`, please?

pankajastro replied:
When the cache profile is enabled 
local mac used_cwd: `PosixPath('/var/folders/t4/sk99_rj509127895hqk1m9x00000gn/T/cosmos/profile/470f1452ea88562203ec89a49bab7e2c6043ef3b7f93e61ce30772e24aa38289')`
CI used_cwd: `PosixPath('/tmp/cosmos/profile/470f1452ea88562203ec89a49bab7e2c6043ef3b7f93e61ce30772e24aa38289'`

tatiana replied:
As we discussed, we've changed to `Path(mock_popen.call_args[0][0][5])`

tatiana reviewed (approved):
Looks great, thanks for addressing all the feedback, @pankajastro !
